flatdata-cpp/include/flatdata/TarFileResourceStorage.h

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2021 HERE Europe B.V.
+ * See the LICENSE file in the root of this project for license details.
+ */
+
+#pragma once
+
+#include "MemoryMappedTarFileStorage.h"
+#include "ResourceStorage.h"
+
+#include <boost/filesystem.hpp>
+
+#include <fstream>
+
+namespace flatdata
+{
+/**
+ * @brief Read-only resource storage for reading flatdata archives inside a TAR file.
+ */
+class TarFileResourceStorage : public ResourceStorage
+{
+public:
+    /**
+     * @brief Create resource storage for a TAR file
+     * @param tar_path The path to the TAR file
+     * @param tar_path The path inside the TAR file
+     * @throws std::runtime_error if the reading of the TAR file fails
+     * @return TarFileResourceStorage or nullptr on error
+     */
+    static std::unique_ptr< TarFileResourceStorage > create( const char* tar_path,
+                                                             const char* sub_path = "" );
+
+    std::unique_ptr< ResourceStorage > create_directory( const char* key ) override;
+    std::unique_ptr< ResourceStorage > directory( const char* key ) override;
+    bool exists( const char* key ) override;
+
+protected:
+    std::shared_ptr< std::ostream > create_output_stream( const char* key ) override;
+    MemoryDescriptor read_resource( const char* key ) override;
+
+private:
+    TarFileResourceStorage( std::shared_ptr< const MemoryMappedTarFileStorage > storage,
+                            const std::string& tar_path,
+                            const std::string& sub_path );
+    std::string get_path( const char* key ) const;
+
+private:
+    std::shared_ptr< const MemoryMappedTarFileStorage > m_storage;
+    std::string m_tar_path;
+    std::string m_sub_path;
+};
+
+// -------------------------------------------------------------------------------------------------
+
+inline std::unique_ptr< TarFileResourceStorage >
+TarFileResourceStorage::create( const char* tar_path, const char* sub_path )
+{
+    std::shared_ptr< const MemoryMappedTarFileStorage > storage(
+        new MemoryMappedTarFileStorage( tar_path ) );
+
+    return std::unique_ptr< TarFileResourceStorage >(
+        new TarFileResourceStorage( storage, tar_path, sub_path ) );
```

Comment:
What about error checking here (e.g. compare to FileResourceStorage which checks some things in `create`)

Reply:
I would not mind printing to std::cerr until we figure out how to properly propagate error messages.